src/identity/interaction/email-password/storage/KeyValueEmailPasswordStore.ts

```diff
@@ -0,0 +1,122 @@
+  webId: string;
+  email: string;
+  password: string;
+}
+
+/**
+ * A payload to persist the fact that a user
+ * has requested to reset their password
+ */
+export interface EmailPasswordForgotPasswordPayload {
+  email: string;
+  recordId: string;
+}
+
+export type EmailPasswordData = EmailPasswordAccountPayload | EmailPasswordForgotPasswordPayload;
+
+export interface ResourceStoreEmailPasswordStoreArgs {
+  baseUrl: string;
+  storagePathname: string;
+  storage: KeyValueStorage<ResourceIdentifier, EmailPasswordData>;
+  saltRounds: number;
+}
+
+/**
+ * A EmailPasswordStore that uses a KeyValueStorage
+ * to persist its information.
+ */
+export class KeyValueEmailPasswordStore extends EmailPasswordStore {
+  private readonly baseUrl: string;
+  private readonly storage: KeyValueStorage<ResourceIdentifier, EmailPasswordData>;
+  private readonly saltRounds: number;
+
+  public constructor(args: ResourceStoreEmailPasswordStoreArgs) {
+    super();
+    this.baseUrl = `${trimTrailingSlashes(args.baseUrl)}${
+      args.storagePathname
+    }`;
+    this.storage = args.storage;
+    this.saltRounds = args.saltRounds;
+  }
+
+  private getAccountResourceIdentifier(key: string): ResourceIdentifier {
+    return { path: `${this.baseUrl}/account/${encodeURIComponent(key)}` };
+  }
+
+  private getForgotPasswordRecordResourceIdentifier(key: string): ResourceIdentifier {
+    return { path: `${this.baseUrl}/forgot-password-resource-identifier/${encodeURIComponent(key)}` };
+  }
+
+  /**
+   * Helper function that converts the given e-mail to an account identifier
+   * and retrieves the account data from the internal storage.
+   */
+  private async getAccountPayload(email: string):
+  Promise<{ identifier: ResourceIdentifier; account?: EmailPasswordAccountPayload }> {
+    const identifier = this.getAccountResourceIdentifier(email);
+    const account = await this.storage.get(identifier) as EmailPasswordAccountPayload | undefined;
+    return { identifier, account };
+  }
+
+  public async authenticate(email: string, password: string): Promise<string> {
+    const { account } = await this.getAccountPayload(email);
+    assert(account, 'No account by that email');
+    assert(await compare(password, account.password), 'Incorrect password');
+    return account.webId;
+  }
+
+  public async create(email: string, webId: string, password: string): Promise<void> {
+    const { identifier, account } = await this.getAccountPayload(email);
+    assert(!account, 'Account already exists');
+    const payload: EmailPasswordAccountPayload = {
+      email,
+      webId,
+      password: await hash(password, this.saltRounds),
+    };
+    await this.storage.set(identifier, payload);
+  }
+
+  public async changePassword(email: string, password: string): Promise<void> {
+    const { identifier, account } = await this.getAccountPayload(email);
+    assert(account, 'Account does not exist');
+    account.password = await hash(password, this.saltRounds);
```

Comment:
Instead passing `saltRounds`, shall we just pass in the hash function?

Reply:
I'm guessing it was done like this because passing a function is not possible in Components.js?